k2/csrc/fsa.h

```diff
@@ -13,7 +13,7 @@
 namespace k2 {
 
 using Label = int32_t;
-using StateId = int32_t;
+using StateId = std::size_t;
```

Comment:
For compactness I want this to stay int32_t. We may at some point want to use -1 for "no such state" which is why I prefer signed.

Reply:
but for now, we use StateId as the index of `fsa.leaving_arcs`, so if we do `state_id < fsa.leaving_arcs.size()`, actually we are comparing in32_t with size_t.

Reply:
static_cast<int32_t> can be used to suppress the warning, though it needs more typing.

Reply:
Another way is that we return type StateId in https://github.com/danpovey/k2/blob/34e90886f278e339f842ca802a172c0f33602ae5/k2/csrc/fsa.h#L77

and use it anywhere required as size.

what do you think?

Reply:
That would be OK I guess.  I don't want to go too far in the OpenFst style of using lots of different equivalent types though.   BTW, I prefer not to use NumStates() in a loop because I dont know how much compilers will optimize it (these days, due to multi-threaded code, compilers can be cautious, I believe).  Better to declare a variable.

Reply:
mm..so we will write such code

```
StateId num_states = fsa.NumStates();
for(; state_id < num_states;)
```
is that ok?
> I prefer not to use NumStates() in a loop because I dont know how much compilers will optimize it

Reply:
Yes thats what I meant but space after `for`.  If someone like @kkm assures me this is not necessary, I am OK to change it.

Reply:
OK. Updated. Please check

Reply:
> BTW, I prefer not to use NumStates() in a loop because I dont know how much compilers will optimize it 

It's tricky. First of all, NumStates() should be marked const. Anyway, const methods like `std::vector<T>::size()` can still change global state, which is why the compiler cannot just optimize away something like NumStates() without seeing all the code being called. In this case, since std::vector is a template class, the compiler can see all the code, so I wouldn't be surprised if the compiler could optimize out the function call. However, it depends upon doing two levels of inlining, and *then* detecting that the number of states is not dependent upon the loop variable to be able to hoist the NumStates() function call above the loop.

The quickest way to investigate these sorts of questions is to look at: https://godbolt.org/ You can see the assembly output of multiple compilers!

But anyway, I would just make the variable outside the for loop like Dan says. It's just not worth it to do all of this mental reasoning. Haha.